RFCs/0070-Fn::Length.md

```diff
@@ -0,0 +1,106 @@
+# RFC for Fn::Length
+
+* **Original Author(s):**: @juegong2
+* **Tracking Issue**: https://github.com/aws-cloudformation/cfn-language-discussion/issues/70
+
+# Summary
+
+We will support an intrinsic function called `Fn::Length` that will return the number of elements in a given list.
+
+# Motivation
+
+CloudFormation users may want to get the number of elements in a given list to set certain resource parameters or generate conditions. It may also help to check if index is out of bound for !Select function. This RFC proposes to support the use case of getting the number of elements of a given list.
+
+* Reference:
+  * https://github.com/aws-cloudformation/cfn-language-discussion/issues/61
+
+# Example(s)
+
+Use `Ref` on a Parameter of type `CommaDelimitedList` or `List<Number>`
+
+```json
+{
+  "Parameters": {
+    "InstanceTypes": {
+      "Type": "CommaDelimitedList"
+    },
+    "NumberList": {
+      "Type": "List<Number>"
+    }
+  }
+  "Conditions": {
+    "Has3InstanceTypes": {
+      "Fn::Equals": [
+        {
+          "Fn::Length": {"Ref": "InstanceTypes"}
+        },
+        3
+      ]
+    },
+    "Has2NumbersInList": {
+      "Fn::Equals": [
+        {
+          "Fn::Length": {"Ref": "NumberList"}
+        },
+        2
+      ]
+    }
+  }
+}
+```
+
+Use a hardcoded list
+```json
+{
+  "Conditions": {
+    "Has3InstanceTypes": {
+      "Fn::Equals": [
+        {
+          "Fn::Length": [ "m5.large", "m5.xlarge" ]
+        },
+        3
+      ]
+    }
+  }
+}
+```
+
+# Limitation
+
+`Fn::Length` will not work with lists that are dynamically generated at provisinoing time. For example, assuming MyBucket is the logical id of a AWS::S3::Bucket resource, the following is not supported:
+```json
+{
+  "Fn::Length": { "Fn::Split": [":", { "Fn::GetAtt": [ "MyBucket", "Arn" ] } ] }
+}
+```
+because the ARN of MyBucket is not known until during provisioning.
+
+**Note:** This is a short-term limitation due to underlying implementation constraints. In the fullness of time, this limitation should be removed.
```

Comment:
I would remove this note, or be very careful about wording. Promising a short-term solution and not meeting this self-imposed deadline is a good way to lose trust.

Additionally, getting  this to be useful during provisioning looks complex from the outside:

Having !Length work during provisioning does seem  similar to how other functions work (so short term might be okay here)
However, most use cases will expect to be able to use an !If with !Length, and this would be a major change in how CloudFormation works. Currently in the If documentation, there is this remark.

> You can only reference other conditions and values from the Parameters and Mappings sections of a template. For example, you can reference a value from an input parameter, but you can't reference the logical ID of a resource in a condition.

Repeating that in the limitations here might be a good idea too

Reply:
Thanks for your feedback! I have updated the wording of this note.

Reply:
👍🏻